middlewares/replace_path.go

```diff
@@ -0,0 +1,21 @@
+package middlewares
+
+import (
+	"net/http"
+)
+
+// ReplacePath is a middleware used to replace the path of a URL request
+type ReplacePath struct {
+	Handler http.Handler
+	Path    string
+}
+
+// ReplacedPathHeader is the default header to set the old path to
+const ReplacedPathHeader = "X-Replaced-Path"
+
+func (s *ReplacePath) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r.Header.Add(ReplacedPathHeader, r.URL.Path)
+	r.URL.Path = s.Path
+	r.RequestURI = r.URL.RequestURI()
```

Comment:
I'm not too deep into this part of HTTP, but is it legitimate to change this?

This is what [net/http](https://golang.org/pkg/net/http/#Request) says:

```
// RequestURI is the unmodified Request-URI of the
// Request-Line (RFC 2616, Section 5.1) as sent by the client
// to a server. Usually the URL field should be used instead.
// It is an error to set this field in an HTTP client request.
RequestURI string
```

WDYT?